### DISCUSSION TITLE
Is there any backward compatibility for Experimental APIs in Airflow stable API [Airflow version 2.0.0 and above]

### DISCUSSION THREAD

**Discussion by achoudha-tibco** [apache/airflow#17460] [Q&A]
Hi team,

We noticed that there are lots of differences between Airflow experimental apis and airflow stable rest apis in terms of resource path and parameters apart from base url.

Since stable apis are available only for airflow version 2.0.0 and higher, is it backward compatible i.e. can we still use the /api/experimental on Airflow version 2.0.0 +?
Reason:
We are in POC to use airflow rest apis to perform operations on dags like list dag run , trigger dags etc. Few of the customers can be on older version of airflow and few on new.  So we don't want to configure 2 versions of client to use these apis.

Second, the airflow we are using is running on GCP cloud composer which is not allowing us to set backed auth to basic_auth.
We tried with backend auth : default and it worked fine.

**Comment by ephraimbuddy** [comment]
Yes. You can still use the experimental API in 2.0.0+

**Reply by potiuk** [reply]
However definitely the experimental/deprecated  API will go away and will be removed in Airflow 3 (and no new features will be added to it any more in Airflow 2).

There is this Migration guide how to migrate from experimental API to the stable one: https://airflow.apache.org/docs/apache-airflow/stable/upgrading-to-2.html#migration-guide-from-experimental-api-to-stable-api-v1

Rather than only using experimental/deprecated API you might actually want to follow the migration guide and support both APIs for a while once you have customers using both 1.10 and 2 version of Airflow. You could have switch to choose which API to use and turn it on by default for all users of Airlfow 2- this way you can fix all problems earlier and then migrate your users to the new API smoothly - with the possibility of switching them back to the old API:

By doing it you have a chance to smoothen your migration - and rather than worrying about a lot of things at the same time in the future you can solve this problem now by investing in API switch.  It's less risky, future-proof and you can even implement it gradually now rather than "big-bang" in the future.

For me it's no brainer and you should use the opportunity now that you can freely switch between those two versions - this opportunity will go away (and sooner than you think - Airflow 3 will be out some time early next year I think).

**Reply by achoudha-tibco** [reply]
@potiuk  Thanks for pointing out about the future roadmap for experimental apis . As rightly said there would be some trade off but seems like supporting both version for the time being will be the way to go.
